byron/ledger/executable-spec/test/Ledger/HasTypeReps/Properties.hs

```diff
@@ -0,0 +1,170 @@
+--   'typeReps' for all types within 'TxIn'.
+exampleTypeRepsTxIn :: Assertion
+exampleTypeRepsTxIn =
+  let txIn = TxIn aTxId 0
+  in (typeReps txIn) @?= typeOf (undefined::TxIn)
+                         <| typeOf (undefined::TxId)
+                         <| typeOf (undefined::Hash)
+                         <| typeOf (undefined::Int)
+                         <| typeOf (undefined::Natural)
+                         <| empty
+
+-- | A 'TxWits' term may contain multiple inputs/outputs/witnesses.
+--   In this example, we have 2 inputs and show how the 'typeReps' for
+--   'TxIn' is repeated twice.
+exampleTypeRepsTx :: Assertion
+exampleTypeRepsTx =
+  let
+    (in0,in1) = (TxIn aTxId 0, TxIn aTxId 1)
+    outs = []
+    wits = []
+    tx = TxWits (Tx [in0, in1] outs) wits
+  in (typeReps tx) @?= typeOf (undefined::TxWits)
+                       <| typeOf (undefined::Tx)
+                       <| typeOf (undefined::[TxIn])
+                       <| typeReps in0
+                       >< typeReps in1
+                       >< (Seq.fromList [ typeOf (undefined::[TxOut])
+                                        , typeOf (undefined::[Wit]) ])
+
+--------------------------------------------------------------------------------
+-- Properties of abstractSize of TxWits / TxIn /TxOut / Wit
+--------------------------------------------------------------------------------
+
+-- | Since 'VKey' appears in both Wit and TxOut (as part of Addr), we
+--   need to use the same "cost" for this type in the following tests.
+vKeyCost :: Size
+vKeyCost = 17
+
+-- | Example costs for types in a 'TxIn'.
+costsTxIn :: Map TypeRep Size
+costsTxIn = Map.fromList [ (typeOf (undefined :: TxIn), 2)
+                         , (typeOf (undefined :: TxId), 3)
+                         , (typeOf (undefined :: Hash), 5) ]
+
+-- | Example costs for types in a 'TxOut'.
+costsTxOut :: Map TypeRep Size
+costsTxOut = Map.fromList [ (typeOf (undefined :: TxOut), 11)
+                          , (typeOf (undefined :: Addr), 13)
+                          , (typeOf (undefined :: VKey), vKeyCost)
+                          , (typeOf (undefined :: Lovelace), 23) ]
+
+-- | Example costs for types in a 'Wit'.
+costsWit :: Map TypeRep Size
+costsWit = Map.fromList [ (typeOf (undefined :: Wit), 31)
+                        , (typeOf (undefined :: VKey), vKeyCost)
+                        , (typeOf (undefined :: Sig Tx), 41)]
+
+-- | This property tests that 'abstractSize' accounts for multiple occurences
+--   of any type represented in the given "cost map".
+--
+--   'abstractSize' uses a "cost map" to compute the size of a term.
+--   For types that contain multiple occurrences of other types, 'abstractSize'
+--   should count each occurence (e.g. a 'Tx' may contain multiple 'TxIn' and
+--   when computing the size of Tx, we should account for the size of each
+--   'TxIn' occurence).
+--
+--  Precondition: this test assumes that the "cost map" only accounts for types
```

Comment:
I have trouble reading this, but maybe because I'm slow :/ In particular what do you mean by:

> the "cost map" only accounts for types that occur within the "inner" type of multiples that we are targeting